pkg/query-service/app/logs/v3/query_builder.go

```diff
@@ -0,0 +1,326 @@
+			}
+			fmtVal := utils.ClickHouseFormattedValue(toFormat)
+
+			if operator, ok := logsOperatorMappingV3[op]; ok {
+				conditions = append(conditions, fmt.Sprintf("%s %s %s", columnName, operator, fmtVal))
+			} else {
+				return "", fmt.Errorf("unsupported operation")
+			}
+		}
+	}
+	queryString := strings.Join(conditions, " AND ")
+
+	if len(queryString) > 0 {
+		queryString = " AND " + queryString
+	}
+	return queryString, nil
+}
+
+func buildLogsQuery(start, end, step int64, mq *v3.BuilderQuery, tableName string, fields map[string]v3.AttributeKey) (string, error) {
+
+	filterSubQuery, err := buildLogsTimeSeriesFilterQuery(mq.Filters, fields)
+	if err != nil {
+		return "", err
+	}
+
+	samplesTableTimeFilter := fmt.Sprintf("(timestamp >= %d AND timestamp <= %d)", start, end)
+
+	selectLabels, err := getSelectLabels(mq.AggregateOperator, mq.GroupBy, fields)
+	if err != nil {
+		return "", err
+	}
+
+	queryTmpl :=
+		"SELECT toStartOfInterval(fromUnixTimestamp64Nano(timestamp), INTERVAL %d SECOND) AS ts" + selectLabels +
+			", %s as value " +
+			"from signoz_logs.distributed_logs " +
+			"where " + samplesTableTimeFilter + "%s " +
+			"group by %s " +
+			"order by %sts"
+
+	groupBy := groupByAttributeKeyTags(mq.GroupBy...)
+	orderBy := orderByAttributeKeyTags(mq.OrderBy, mq.GroupBy)
+
+	aggregationKey := ""
+	if mq.AggregateAttribute.Key != "" {
+		aggregationKey, err = getClickhouseColumnName(mq.AggregateAttribute, fields)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	switch mq.AggregateOperator {
+	case v3.AggregateOperatorRate:
+		op := fmt.Sprintf("count(%s)/%d", aggregationKey, step)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, orderBy)
+		return query, nil
+	case
+		v3.AggregateOperatorRateSum,
+		v3.AggregateOperatorRateMax,
+		v3.AggregateOperatorRateAvg,
+		v3.AggregateOperatorRateMin:
+		op := fmt.Sprintf("%s(%s)/%d", aggregateOperatorToSQLFunc[mq.AggregateOperator], aggregationKey, step)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, orderBy)
+		return query, nil
+	case
+		v3.AggregateOperatorP05,
+		v3.AggregateOperatorP10,
+		v3.AggregateOperatorP20,
+		v3.AggregateOperatorP25,
+		v3.AggregateOperatorP50,
+		v3.AggregateOperatorP75,
+		v3.AggregateOperatorP90,
+		v3.AggregateOperatorP95,
+		v3.AggregateOperatorP99:
+		op := fmt.Sprintf("quantile(%v)(%s)", aggregateOperatorToPercentile[mq.AggregateOperator], aggregationKey)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, orderBy)
+		return query, nil
+	case v3.AggregateOperatorAvg, v3.AggregateOperatorSum, v3.AggregateOperatorMin, v3.AggregateOperatorMax:
+		op := fmt.Sprintf("%s(%s)", aggregateOperatorToSQLFunc[mq.AggregateOperator], aggregationKey)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, orderBy)
+		return query, nil
+	case v3.AggregateOpeatorCount:
+		op := "toFloat64(count(*))"
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, orderBy)
```

Comment:
If they provide an aggregate key for `count`, they mean that it exists. Is that honoured? For instance, count of `passthrough.this.codepath` by service. Here I am trying to count the number of records seen for each interval with `passthrough.this.codepath`. It appears to me that currently, it counts all regardless of the existence of `passthrough.this.codepath` which is not the expected behaviour.

Reply:
Have added a `has` clause.